src/cli/serve/legacy_config.js

```diff
@@ -0,0 +1,47 @@
+import { noop, transform } from 'lodash';
+
+// legacySettings allow kibana 4.2+ to accept the same config file that people
+// used for kibana 4.0 and 4.1. These settings are transformed to their modern
+// equivalents at the very begining of the process
+export const legacySettings = {
+  // server
+  port: 'server.port',
+  host: 'server.host',
+  pid_file: 'pid.file',
+  ssl_cert_file: 'server.ssl.cert',
+  ssl_key_file: 'server.ssl.key',
+
+  // logging
+  log_file: 'logging.dest',
+
+  // kibana
+  kibana_index: 'kibana.index',
+  default_app_id: 'kibana.defaultAppId',
+
+  // es
+  ca: 'elasticsearch.ssl.ca',
+  elasticsearch_preserve_host: 'elasticsearch.preserveHost',
+  elasticsearch_url: 'elasticsearch.url',
+  kibana_elasticsearch_client_crt: 'elasticsearch.ssl.cert',
+  kibana_elasticsearch_client_key: 'elasticsearch.ssl.key',
+  kibana_elasticsearch_password: 'elasticsearch.password',
+  kibana_elasticsearch_username: 'elasticsearch.username',
+  ping_timeout: 'elasticsearch.pingTimeout',
+  request_timeout: 'elasticsearch.requestTimeout',
+  shard_timeout: 'elasticsearch.shardTimeout',
+  startup_timeout: 'elasticsearch.startupTimeout',
+  verify_ssl: 'elasticsearch.ssl.verify',
+};
+
+// transform legacy options into new namespaced versions
+export function rewriteLegacyConfig(object, log = noop) {
+  return transform(object, (clone, val, key) => {
+    if (legacySettings.hasOwnProperty(key)) {
```

Comment:
Now that I'm implementing this, and tracking the keys in their own variable, I'm really confused how this is beneficial?

Is `hasOwnProperty()` too long of a method? What do you find beneficial about `keys().includes()`? If this is about aesthetics then I think that `Map` exposes the best API -- [comparison](https://gist.github.com/spalger/83334179a40215a5dc08aaf6520e7e82)

If it's about complexity, then counter that tracking the key list of an outside of the object just so you can reimplement `hasOwnProperty()` is quite unexpected. It would be more confusing to me (and most I would assuming) since `hasOwnProperty()` is an old and well understood method that already does this.